### DISCUSSION TITLE
Can I use my own Auth Server to forward browser requests to Ory Kratos?

### DISCUSSION THREAD

**Discussion by tylerku** [ory/kratos#1314] [Q&A] [answered]
I want to have the setup like this (Broswer site -> Auth Server -> Ory Kratos) but it seems like Kratos is not designed to work this way? I've been looking at the registration endpoints specifically. I read the big red warning on the documentation that says not to use `/self-service/registration/api` or `/self-service/registration?flow=` from a nodejs backend app cuz of security vulnerabilities (CSRF). So, am I just out of luck if I want to use my own microservice on top of ory kratos? 

For more background, the reason we want to have our own auth service is so we can make a JWT and pass it back to our browser application. Then, we can have our browser app use that jwt for auth when hitting our api.. and our api wont have to make another request to ory kratos to check the session because it can verify the jwt on its own.

Thanks for your help!

**Comment by brycelund** [comment, marked as answer]
We're in the same boat. We're looking for something that is essentially an open source micro service replacement for Amazon Cognito. Can Ory Kratos be used in that capacity? As I understand it now, in order to sign up a user you have to call `/self-service/registration/api` to get a flow id, and then call `/self-service/registration/methods/password?flow=` to create a user. Is there any way to do this from a backend in only one request? The Admin API doesn't allow setting a password when registering a user.

**Reply by aeneasr** [reply]
It is not a good idea to work around the browser flows if you target browser applications. We spent a lot of time and effort to make these flows secure and by just doing your own thing here you risk some serious vulnerabilities.

If you still wish to use the API flows - there is currently no way to work around the initialization flow. We might fix that in a future version but I don't think that it is prohibitive to the use? It's just one more call :)

Regarding JWTs - they are not a good match for session storage. If you log users out, the JWT will be valid for as long as you put the JWT to be valid for. That might be - depending on how long you keep users signed in - several weeks. I highly encourage you **not to do this**.

If you want microservice JWTs you can use Ory Oathkeeper to convert the session cookie to a JWT which you can use for a short period of time (e.g. 5 minutes) in your backend services.

In general, trying to work around the security mechanisms from Ory Kratos is a really, really bad idea and **you will expose your data, your users, your application, and your business to serious risks.**

**Reply by tylerku** [reply]
So if I understand right we should just not run an auth service and have our web app make calls directly to our Ory Kratos instance, and then just not use JWTs and if we really want to we can convert the session cookie to a JWT for a brief period.  Assuming we don’t use a JWT then the idea is when our web app makes calls to our other microservices, those services will check the session id with our ory kratos instance to make sure the user is signed in. Does that sound right?

**Reply by aeneasr** [reply]
Yes, exactly. If you have many services calling Kratos you can replace that with an initial request to Kratos at the API Gateway/Reverse Proxy which converts it to a JWT for your internal infrastructure

**Comment by kawtharmahboubi** [comment]
Our use case is similar to your use case , I was wondering @aeneasr , if we use our auth api to redirect kratos response to the frontend would it be still a problem ? 
We need our auth api to do some specific logic on top of kratos , so if we follow Kratos logic/flows but initialized with our backend would it be still a problem ?

**Reply by jonnyijapan** [reply]
I think I'm about to do the same thing as you all describe here.

And to create an identity using only the "backend" > "kratos", I currently use this admin API, it works well
https://www.ory.sh/docs/kratos/reference/api#operation/adminCreateIdentity

Unless I understand your question wrong.

Here is an example query, exported from Postman to cURL command. The domain/port here is the private admin one.

```
curl --location --request POST 'http://127.0.0.1:4434/admin/identities' \
--header 'Content-Type: application/json' \
--data-raw '{
    "schema_id": "default",
    "traits": {
        "email": "foo@bar.net"
    },
    "credentials": {
        "password": {
            "config": {
                "password": "secretPassWord"
            }
        }
    }
}'
```
